libs/ngx-sherfire/src/lib/firestore/derivable-firestore.service.ts

```diff
@@ -0,0 +1,136 @@
+        return v instanceof DocumentReference
+            ? this.CACHES.doc.data(v)
+            : v instanceof CollectionReference
+            ? this.CACHES.collection.data(v)
+            : this.queryData$(v);
+    }
+
+    private readonly CACHES = this.buildCaches();
+
+    private querySnapshot$<AppModelType, DbModelType extends DocumentData>(
+        query: Query<AppModelType, DbModelType>,
+    ): Derivable<QuerySnapshot<AppModelType, DbModelType>> {
+        return fromEventPattern<QuerySnapshot<AppModelType, DbModelType>>(value$ =>
+            onSnapshot(query, ...this.snapshotEventHandlers(value$)),
+        );
+    }
+
+    private docSnapshot$<AppModelType, DbModelType extends DocumentData>(
+        ref: DocumentReference<AppModelType, DbModelType>,
+    ): Derivable<DocumentSnapshot<AppModelType, DbModelType>> {
+        return fromEventPattern<DocumentSnapshot<AppModelType, DbModelType>>(value$ =>
+            onSnapshot(ref, ...this.snapshotEventHandlers(value$)),
+        );
+    }
+
+    private snapshotEventHandlers<T>(value$: DerivableAtom<T>) {
+        return [
+            (snap: T) => this.zone.run(() => value$.set(snap)),
+            // istanbul ignore next, because I don't know how to force that to happen
+            (err: FirestoreError) => this.zone.run(() => value$.setFinal(error(err))),
+        ] as const;
+    }
+
+    private queryDocs$<AppModelType, DbModelType extends DocumentData>(
+        query: Query<AppModelType, DbModelType>,
+    ): Derivable<Array<QueryDocumentSnapshot<AppModelType, DbModelType>>> {
+        return this.snapshot$(query).map(s => s.docs);
+    }
+
+    private queryData$<AppModelType, DbModelType extends DocumentData>(
+        query: Query<AppModelType, DbModelType>,
+    ): Derivable<AppModelType[]> {
+        return this.docs$(query).map(snaps => snaps.map(snap => snap.data()));
+    }
+
+    private buildCaches() {
+        const CACHE_OPTIONS: DerivableCacheOptions<
+            DocumentReference<any, any> | CollectionReference<any, any>,
+            string
+        > = {
+            delayedEviction: true,
+            mapKeys: v => v.path,
+        };
+
+        return {
+            collection: {
+                snapshot: derivableCache(
+                    (ref: CollectionReference<any, any>) => this.querySnapshot$(ref),
+                    CACHE_OPTIONS,
+                ),
+                docs: derivableCache((ref: CollectionReference<any, any>) => this.queryDocs$(ref), CACHE_OPTIONS),
+                data: derivableCache((ref: CollectionReference<any, any>) => this.queryData$(ref), CACHE_OPTIONS),
+            },
+            doc: {
+                snapshot: derivableCache((ref: DocumentReference<any, any>) => this.docSnapshot$(ref), CACHE_OPTIONS),
+                data: derivableCache(
+                    (ref: DocumentReference<any, any>) =>
+                        lens({
+                            get: () => this.snapshot$(ref).get().data(),
```

Comment:
Make sure to remove the repeated call to `this.snapshot$` if you decide to remove caching.